# Python Pandas Tutorial
**step:1 Creat a folder on your pc**<br><br>
**step:2 Installing part**
*Windows ke andar apne folder me terminal kholne ke liye shift+right click and the open powershell window after open terminal you have to install these things, pip install numpy, pip install pandas, pip install jupyter and for upgrade pip install --upgrade (numpy)*<br><br>
**step:3 Open jupyter notebook** *for open jupyter notebook on your forlder pehle terminal khole or 'jupyter notebook' type karke enter kar de aapke folder par jupyter notebook khul jayega*<br><br>
**step:4 Create new file on jupyter notebook** *jupyter notebook kholne ke bad apko right top side me new ka option dikh raha hoga usko click karke python 3 me click karna hai apka folder ban jayega fir aap usko rename bhi kar sakte hai*


In [1]:
import numpy as np
import pandas as pd

In [2]:
dict1 = {
    "name":['sanjiv', 'samar', 'shubham', 'nitesh'],
    "marks":[85, 87, 90, 85],
    "city":['jamshedpur', 'ranchi', 'musabani', 'bokaro']
}

In [3]:
df = pd.DataFrame(dict1)

*ye simply ek dataframe bana dega, dataframe is like a excel sheet. ye inko ek excel sheet me convert kar dega faster indexing ke liye, or numpy ke functiuons ko hum use kar payenge.ye sari chije fast karega Big data ko hum ache se analyze kar payenge*

In [4]:
df

name  marks        city
0   sanjiv     85  jamshedpur
1    samar     87      ranchi
2  shubham     90    musabani
3   nitesh     85      bokaro

*or mujhe df ko dekhna hai to kuch is tarah dekh sakta hun,<br><br> isne kya kiya? isne yahan par name de diya marks de diya or city de diya, jaise humne **dict1** me diya tha **name, marks, city** usi tarah isne column ke name de diye hai name, marks or city, usi ke sath-sath jo ye 0,1,2,3 dekh rahe hai ye **index** hai ye **har row ko identifie karta hai index se or column ko column name** se<br><br> isi chij ko isi **dataframe** ko **csv me excelsheet me hum export kaise karenge** abhi dekhenge*

In [5]:
df.to_csv('friends.csv')

*to,  ***to_csv*** function se apka dataframe, csv file me convert ho jayega or fir aap isko excelsheet me bhi khol sakte hai*<br><br>
*to simply kisi bhi data ko analyse karne ke bad aapko usko excelsheet me dalna hai to simply ***to_csv*** function ka istemal kar skte hai aap*

*#--excelsheet me **index** ko hatane ke liye hum kya karenge*<br><br> *---yaniki me cahata hun apni csv me index na dekhu*

In [6]:
df.to_csv('friends_index_false.csv', index=False)

## 1.Dataframe ko hum dekhne ke liye kya kya kar sakte hai

*--Agar mera dataframe bhaut bada hota to like 1-2 millions rows hote agar, to ye sari 1-2 milllions rows to dikah nahi sakta to ye thoda sa dikhata hai.*<br>
**man lo aapko dekhna hai suru ke 3 rows to apko likhna hoga**

In [7]:
df.head(3)

name  marks        city
0   sanjiv     85  jamshedpur
1    samar     87      ranchi
2  shubham     90    musabani

**or agar apko akhir ke 3 rows dekhne hai**

In [8]:
df.tail(3)

name  marks      city
1    samar     87    ranchi
2  shubham     90  musabani
3   nitesh     85    bokaro

***--df.describe--***<br>
*ye kya karega jo bhi **numericals column** hai jaise ki mere kon kon se marks hai. is numericals column ka yahan par **count, mean, standard divison, minimum, 25% percentile** ye sari chije calculate karega aapke liye ek **statistical analysis** karega maximum batayega aapko kitna hoga*

In [9]:
df.describe()

marks
count   4.000000
mean   86.750000
std     2.362908
min    85.000000
25%    85.000000
50%    86.000000
75%    87.750000
max    90.000000

# 2.Import csv

*abb man lo me ek csv banata hun excel me,<br>
**csv: simply coma separated values ki sheet hoti hai**<br>
--abb humne yahan par sanjiv.csv name ka ek excel file banaya hai, jisme humne bs random values dale hai, fri hume isko save kar diya* <br><br>
*abb sanjiv.csv ko read karne ke liye hum sanjiv name ka ek dataframe bana sakte hai*<br>
**dataframe: ek excelsheet hai jo ki python ke andar hum istemal kar sakte hai**


In [10]:
sanjiv = pd.read_csv('sanjiv.csv')

In [11]:
sanjiv

Unnamed: 0  train no  speed        city
0           0      1125     92  jamshedpur
1           1      5847     87      ranchi
2           2      4455     90    musabani
3           3      2365     85      bokaro

## 3.For change csv values

In [12]:
sanjiv['speed']

0    92
1    87
2    90
3    85
Name: speed, dtype: int64

In [13]:
sanjiv['speed'][0]

92

In [14]:
sanjiv['speed'][0]=92

C:\Users\SANJIV~1\AppData\Local\Temp/ipykernel_11692/2498308048.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sanjiv['speed'][0]=92


*to ye bs bol raha aap aise slice par value ko set na kare*<br><br>
**or agar hum fir se sanjiv ko print kare to aap dekhenge ki value jo hai wo change ho gayi hai**

In [15]:
sanjiv

Unnamed: 0  train no  speed        city
0           0      1125     92  jamshedpur
1           1      5847     87      ranchi
2           2      4455     90    musabani
3           3      2365     85      bokaro

**or agar hum isko csv me export kar de to humare excelsheet ka bhi value change ho jayega**

In [16]:
sanjiv.to_csv('sanjiv.csv')

**to is tarah hum value ko change karke wapas csv me export kar sakte hai**

## 4.For change index

In [17]:
sanjiv.index = ['first', 'second', 'third', 'fourth']

In [18]:
sanjiv

Unnamed: 0  train no  speed        city
first            0      1125     92  jamshedpur
second           1      5847     87      ranchi
third            2      4455     90    musabani
fourth           3      2365     85      bokaro

*to jaruri nahi hai ki jo aapki index hai, wo 0,1,2,3 hi hoga ya koyi integer hi rahega, aap yahan par **strings** ka bhi istemal kar sakte hai. aaplog yahan par koi sa bhi number chun sakte hai<br><br>
**bs ye samajhiye ki "row ko access karne ke liye index hota hai or column ko access karne ke liye column ka name hota hai"**

# 5.what is pandas:

**1.Pandas is an open sources data analysis library written in python**<br>
**2.It leverages the power and speed of numpy to make data analysis and preprocessing easy for data scientist.**<br>
**3.It provides rich and highly robust data operations.**<br>
# 6.Pandas data structure 
### Pandas has two types of data structures:
*a.**series-** it's a one dimensional array with indexes, it stores a single column or row of data in a DataFrame.*<br>
*b.**DataFrame-** it's tabular spreadsheet like structure representing rows each of which contains one or multiple columns*<br>
*--A one-dimensional array (labeled) capable of holding any type of data - **series** .*<br>
*--A two-dimensional data (labeled) structure with columns of potentially different type of data - **DataFrame** .*

In [19]:
df

name  marks        city
0   sanjiv     85  jamshedpur
1    samar     87      ranchi
2  shubham     90    musabani
3   nitesh     85      bokaro

In [20]:
type(df['marks'])

pandas.core.series.Series

**hum yahan sekh sakte hai jo column marks hai wo ek "series" hai**

## Now we'll create a new dataframe, and from here we'll look at the difference between the series and the dataframe, and we'll explore everything here.


# 1.Series
### Series ek data structure hota hai pandas ka, jo ki pandas ki ek basic index hoti hai 

**random series ko print karne ke liye**

In [21]:
ser = pd.Series(np.random.rand(34))

In [22]:
ser

0     0.091403
1     0.738452
2     0.435066
3     0.148843
4     0.041055
5     0.704245
6     0.724621
7     0.108592
8     0.349322
9     0.716949
10    0.805213
11    0.438844
12    0.869548
13    0.830362
14    0.181636
15    0.588682
16    0.253745
17    0.339227
18    0.581635
19    0.585920
20    0.655013
21    0.998558
22    0.641631
23    0.615030
24    0.852233
25    0.285600
26    0.494614
27    0.967041
28    0.197912
29    0.218176
30    0.630686
31    0.688535
32    0.925282
33    0.364969
dtype: float64

**to isne mujhe 0 se 33 tak ek series de diya**<br>
**agar me iska type dekhna chahta hun to-- ye mujhe batayega ye ek series hai<br>** ***Note: Series ka 'S' bada hona cahiye***

In [23]:
type(ser)

pandas.core.series.Series

# 2.DataFrame

In [24]:
newdf = pd.DataFrame(np.random.rand(334, 5), index=np.arange(334))

In [25]:
newdf.head()

0         1         2         3         4
0  0.410182  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

*ye mujhe suru ke 5 rows dikhayega*

In [26]:
newdf.tail()

0         1         2         3         4
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

*ye mujhe akhir ke 5 rows dikhayega*

In [27]:
newdf

0         1         2         3         4
0    0.410182  0.735723  0.351642  0.316163  0.028390
1    0.490249  0.464304  0.337367  0.814775  0.299901
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
4    0.840038  0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...       ...
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

[334 rows x 5 columns]

*agar me isme kuch na lagaun to ye mujhe sari ki sari rows dikha diya 4 ke bad dot dot or fir 329 se dikhaya ***or ye mujhe bata raha hai ki apke dataframe me 334 rows or 5 columns hai*** .*

In [28]:
type(newdf)

pandas.core.frame.DataFrame

*or iska type check karne se ye bata raha hai ki ye ek ***DataFrame*** hai*

In [29]:
newdf.head()

0         1         2         3         4
0  0.410182  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

In [30]:
newdf.describe()

0           1           2           3           4
count  334.000000  334.000000  334.000000  334.000000  334.000000
mean     0.514542    0.524246    0.518095    0.492071    0.498725
std      0.285607    0.281581    0.288812    0.288791    0.284334
min      0.002447    0.003221    0.006087    0.000677    0.001619
25%      0.266530    0.291856    0.275958    0.225955    0.253616
50%      0.524903    0.544120    0.498824    0.486123    0.502969
75%      0.770232    0.755031    0.758125    0.738512    0.742498
max      0.996860    0.995377    0.996208    0.999514    0.993394

In [31]:
newdf.dtypes

0    float64
1    float64
2    float64
3    float64
4    float64
dtype: object

*ye hume hume bata raha hai kya kya data type hai humare columns ke*

In [32]:
newdf[0][0]='sanjiv'

In [33]:
newdf.dtypes

0     object
1    float64
2    float64
3    float64
4    float64
dtype: object

*to aap dekh sakte hai humne iska data type change kar diya column 0 par abb ye ***object*** dikha raha hai kyuki humne yahan par ek ***string*** dal diya hai*

In [34]:
newdf.head()

0         1         2         3         4
0    sanjiv  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

*dekh sakte hai humne yahan par ek string dal diya hai*

In [35]:
newdf.index

Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            324, 325, 326, 327, 328, 329, 330, 331, 332, 333],
           dtype='int64', length=334)

*to ye hume humara indexc bata raha hai jo 0 se 333 ja hai or iska length hai 334*

In [36]:
newdf.columns

RangeIndex(start=0, stop=5, step=1)

*to ye bata hai 0 se 5 tak humara columns hai*

## 3.newdf DataFrame ko numpy array me convert karenge

In [37]:
newdf.to_numpy()

array([['sanjiv', 0.735723091454385, 0.3516424862501931,
        0.31616291048002754, 0.02838993699248249],
       [0.49024859145822886, 0.46430407379111593, 0.33736670619461195,
        0.8147754528745774, 0.2999007651510046],
       [0.26226696923976545, 0.9376193352594667, 0.8315238265145201,
        0.11802968568817018, 0.5252062225648687],
       ...,
       [0.117436859151556, 0.4604396750677855, 0.8765619304054042,
        0.636200826584253, 0.22338142209918777],
       [0.381840744234929, 0.5665916447538092, 0.37839918472465905,
        0.803675505975083, 0.020226039105047278],
       [0.8378166049758905, 0.8303186345675537, 0.09782367544275561,
        0.1376419992721637, 0.10283467875826702]], dtype=object)

*to isne newdf dataframe ko numpy array me conver kar diya ***par isko iska dtype object hi rakhna pada kyuki hume yaha 0th column par ek string ko dala hai*** .* 

In [38]:
newdf[0][0]=0.1

In [39]:
newdf.head()

0         1         2         3         4
0       0.1  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

In [40]:
newdf.to_numpy()

array([[0.1, 0.735723091454385, 0.3516424862501931, 0.31616291048002754,
        0.02838993699248249],
       [0.49024859145822886, 0.46430407379111593, 0.33736670619461195,
        0.8147754528745774, 0.2999007651510046],
       [0.26226696923976545, 0.9376193352594667, 0.8315238265145201,
        0.11802968568817018, 0.5252062225648687],
       ...,
       [0.117436859151556, 0.4604396750677855, 0.8765619304054042,
        0.636200826584253, 0.22338142209918777],
       [0.381840744234929, 0.5665916447538092, 0.37839918472465905,
        0.803675505975083, 0.020226039105047278],
       [0.8378166049758905, 0.8303186345675537, 0.09782367544275561,
        0.1376419992721637, 0.10283467875826702]], dtype=object)

# 4.Attributes

In [41]:
newdf.T

0         1         2         3         4         5         6    \
0       0.1  0.490249  0.262267  0.287334  0.840038  0.154071  0.433174   
1  0.735723  0.464304  0.937619  0.974052  0.276021  0.032365  0.140683   
2  0.351642  0.337367  0.831524  0.944044  0.140503  0.077466  0.714506   
3  0.316163  0.814775   0.11803   0.86909  0.599835   0.50315  0.572959   
4   0.02839  0.299901  0.525206  0.136935   0.51121  0.778425  0.167123   

        7         8         9    ...       324       325       326       327  \
0  0.801083  0.338782  0.117987  ...  0.468199   0.77043   0.29094  0.943797   
1    0.5863  0.177624   0.47847  ...  0.259284  0.550451  0.717199  0.840112   
2  0.645051  0.225024  0.489211  ...  0.675429  0.024828  0.535824  0.985291   
3  0.225731  0.661354  0.171562  ...   0.70052    0.0646  0.960058  0.858793   
4  0.148068  0.671782  0.373822  ...  0.663197  0.600671  0.550297  0.809531   

        328       329       330       331       332       333  
0  0.031679  0.770738  0.845293  0.117437  0.381841  0.837817  
1  0.597769  0.692331  0.832767   0.46044  0.566592  0.830319  
2  0.337845  0.882043  0.497827  0.876562  0.378399  0.097824  
3   0.38578  0.763866  0.485871  0.636201  0.803676  0.137642  
4  0.153512  0.218966  0.769743  0.223381  0.020226  0.102835  

[5 rows x 334 columns]

*yahan humne dataframe ko ***transpose*** kar diya hai*

### Transpose: means columns row me badal jayenge or row column me(pehle 334 row tha humare pass, abb 334 columns hai or 5 rows hai) like matrix ka transpose

In [42]:
newdf

0         1         2         3         4
0         0.1  0.735723  0.351642  0.316163  0.028390
1    0.490249  0.464304  0.337367  0.814775  0.299901
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
4    0.840038  0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...       ...
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

[334 rows x 5 columns]

In [43]:
newdf.head()

0         1         2         3         4
0       0.1  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

# 5.Sorting

In [44]:
newdf.sort_index(axis=0, ascending=False)

0         1         2         3         4
333  0.837817  0.830319  0.097824  0.137642  0.102835
332  0.381841  0.566592  0.378399  0.803676  0.020226
331  0.117437  0.460440  0.876562  0.636201  0.223381
330  0.845293  0.832767  0.497827  0.485871  0.769743
329  0.770738  0.692331  0.882043  0.763866  0.218966
..        ...       ...       ...       ...       ...
4    0.840038  0.276021  0.140503  0.599835  0.511210
3    0.287334  0.974052  0.944044  0.869090  0.136935
2    0.262267  0.937619  0.831524  0.118030  0.525206
1    0.490249  0.464304  0.337367  0.814775  0.299901
0         0.1  0.735723  0.351642  0.316163  0.028390

[334 rows x 5 columns]

*to ***axis=0 means rows*** to humne yaha rows ko ulta sort kiya hai*

In [45]:
newdf.sort_index(axis=1, ascending=False)

4         3         2         1         0
0    0.028390  0.316163  0.351642  0.735723       0.1
1    0.299901  0.814775  0.337367  0.464304  0.490249
2    0.525206  0.118030  0.831524  0.937619  0.262267
3    0.136935  0.869090  0.944044  0.974052  0.287334
4    0.511210  0.599835  0.140503  0.276021  0.840038
..        ...       ...       ...       ...       ...
329  0.218966  0.763866  0.882043  0.692331  0.770738
330  0.769743  0.485871  0.497827  0.832767  0.845293
331  0.223381  0.636201  0.876562  0.460440  0.117437
332  0.020226  0.803676  0.378399  0.566592  0.381841
333  0.102835  0.137642  0.097824  0.830319  0.837817

[334 rows x 5 columns]

*or ***axis=1 means column*** to humne yahan column ko ulta sort kiya hai*

In [46]:
newdf.head()

0         1         2         3         4
0       0.1  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

In [47]:
newdf[0]

0           0.1
1      0.490249
2      0.262267
3      0.287334
4      0.840038
         ...   
329    0.770738
330    0.845293
331    0.117437
332    0.381841
333    0.837817
Name: 0, Length: 334, dtype: object

In [48]:
type(newdf[0])

pandas.core.series.Series

*to aap dekh sakte hai ye ek series hai ***to series ke combination se dataframe banta hai*** .*

# 6.Copy and View
### (A).view

In [49]:
newdf2 = newdf

*aapko kya lagta hai kya newdf pura ka pura copy ho kar ek naya banega, aisa nahi hoga ***newdf2 jo hai bs ek view hai*** .* 

In [50]:
newdf2[0][0] = 9856 

In [51]:
newdf

0         1         2         3         4
0        9856  0.735723  0.351642  0.316163  0.028390
1    0.490249  0.464304  0.337367  0.814775  0.299901
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
4    0.840038  0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...       ...
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

[334 rows x 5 columns]

*to aap dekh sakte hai jo humne newdf2 me 9856 ko set kiya wo yahan newdf me aa gaya to aisa kyu hua?*<br>**to yahan par hua kya newdf2 newdf ka ek view hai, view ka mtlb "ki wo memmory location me to wohi par hai, bas usko point kar raha hai. Bas ek tarah se ponter ki tarah aap samajh lo*** *newdf2 ko change karne se newdf bhi change ho jayega*  

## (B).Copy

In [52]:
newdf2 = newdf.copy()

*hum copy ke liye .copy ka istemal karte hai jo ki ek dataframe ka function hota hai*

In [53]:
newdf2[0][0] =98562

C:\Users\SANJIV~1\AppData\Local\Temp/ipykernel_11692/501776143.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf2[0][0] =98562


In [54]:
newdf

0         1         2         3         4
0        9856  0.735723  0.351642  0.316163  0.028390
1    0.490249  0.464304  0.337367  0.814775  0.299901
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
4    0.840038  0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...       ...
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

[334 rows x 5 columns]

*to aap dekh sakte hai ab isne newdf me 98562 ko nahi dikhaya kyuki mene yahan par .copy likha*<br><br>***to isiliye dataframe ko equalto karke naya banate hai to copy humesa nahi banti hai jab tak aap .copy ka istemal na kare***<br>*or ye warning kuch usi se related hai*<br><br>**abb hota kya hai ki man lo aap chaining kar rahe ho, aapne bola ki is dataframe ka [0][0] = kuch value kar diya.... to yahan bich me kahan view return karega kahan copy return karega koi guaranty nahi hai. kabhi kabhi ye tarack karna himlog ke liye bhaut muskil ho jata hai ki view return hogi ya copy return hogi.**<br>***kabhi kabhi pandas apni internam memory ke hisab se decide karti hai view return karna hai ki copy return karna hai.***<br><br>**to isiliye hum kabhi kabhi value ko set karte hai to "loc function" ka istemal karte hai**  

# 7.loc function
**hum loc function ko dot loc karke likhte hai**

In [55]:
newdf.loc[0,0] = 852

In [56]:
newdf.head(2)

0         1         2         3         4
0       852  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901

***to aap dikh sakte hai yahan number change ho gaya hai 852 se, or jo yahan column hai wo kuch bhi ho sakta tha ABCDE***<br><br>*to agar columns ke name ko change kar du to*

In [57]:
newdf.columns = list("ABCDE")

In [58]:
newdf.head(2)

A         B         C         D         E
0       852  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901

In [59]:
newdf.loc[0,0] = 852

In [60]:
newdf.head(2)

A         B         C         D         E      0
0       852  0.735723  0.351642  0.316163  0.028390  852.0
1  0.490249  0.464304  0.337367  0.814775  0.299901    NaN

*to yahan aap dekh sakte hai ise [0,0] mila hi nahi to isne ek naya column bana diya [0,0] ka* **to ayiye fir dekhte hai hum in columns ke sath loc function ko kaise use kar sakte hai**

In [61]:
newdf.loc[0,'A'] = 8521

In [62]:
newdf.head(2)

A         B         C         D         E      0
0      8521  0.735723  0.351642  0.316163  0.028390  852.0
1  0.490249  0.464304  0.337367  0.814775  0.299901    NaN

*to yahan aap dekh sathe hai abb isne change kar diya hai number ko to isko address karane ke hume [0,'A'] likhna padha* ***abb humne jo extra column banaya tha hum usko hutayenge***

In [63]:
newdf = newdf.drop(0, axis=1)

In [64]:
newdf.head()

A         B         C         D         E
0      8521  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

*to aap dekh sakte hai jo extra column tha usko humne hata diya hai, ***kuch set of columns dekhne ke liye hum kya kar sakte hai*** .*

In [65]:
newdf.loc[[1,2], ['C', 'D']]

C         D
1  0.337367  0.814775
2  0.831524  0.118030

*par isse humara newdf change nahi hoga "ye bas copy ko return kar raha hai inplace change nahi kar raha hai". usko sirf ye return kar raha hai ek naye dataframe me.*

In [66]:
newdf.head()

A         B         C         D         E
0      8521  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

*dekh sakte hai humara newdf waise ka waise hi hai*<br><br>***sare ke sare rows ko dekhne ke liye hum kya kar sakte hai***

In [67]:
newdf.loc[:, ['C', 'D']]

C         D
0    0.351642  0.316163
1    0.337367  0.814775
2    0.831524  0.118030
3    0.944044  0.869090
4    0.140503  0.599835
..        ...       ...
329  0.882043  0.763866
330  0.497827  0.485871
331  0.876562  0.636201
332  0.378399  0.803676
333  0.097824  0.137642

[334 rows x 2 columns]

***sare ke sare columns ko dekhne ke liye hum kya kar sakte hai***

In [68]:
newdf.loc[[1, 2], :]

A         B         C         D         E
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206

# 8.complex query: searching

In [69]:
newdf.loc[(newdf['A']<0.3)]

A         B         C         D         E
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
5    0.154071  0.032365  0.077466  0.503150  0.778425
9    0.117987  0.478470  0.489211  0.171562  0.373822
12    0.29989  0.943389  0.884019  0.462270  0.748555
..        ...       ...       ...       ...       ...
315  0.017056  0.080788  0.117911  0.176318  0.327976
323   0.24903  0.175113  0.157796  0.568432  0.511504
326   0.29094  0.717199  0.535824  0.960058  0.550297
328  0.031679  0.597769  0.337845  0.385780  0.153512
331  0.117437  0.460440  0.876562  0.636201  0.223381

[93 rows x 5 columns]

*humne kaha jahan jahan par row meri column A par 0.3 se kam ho dikha do,* ***hum isse or complex query bhi search kar sakte hai***

In [70]:
newdf.loc[(newdf['A']<0.3) & (newdf['C']>0.1)]

A         B         C         D         E
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
9    0.117987  0.478470  0.489211  0.171562  0.373822
12    0.29989  0.943389  0.884019  0.462270  0.748555
23   0.201006  0.149415  0.556512  0.045153  0.912707
..        ...       ...       ...       ...       ...
315  0.017056  0.080788  0.117911  0.176318  0.327976
323   0.24903  0.175113  0.157796  0.568432  0.511504
326   0.29094  0.717199  0.535824  0.960058  0.550297
328  0.031679  0.597769  0.337845  0.385780  0.153512
331  0.117437  0.460440  0.876562  0.636201  0.223381

[86 rows x 5 columns]

*to humne yahan ek or condition laga diya hai*

# 9.iloc function
***iloc ka istemal hum tab karte hai jab hume index se count karke value ko target karna hota hai, chahe rows or columns ke name kuch bhi ho***

In [71]:
newdf.head()

A         B         C         D         E
0      8521  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

In [72]:
newdf.iloc[0,4]

0.02838993699248249

*to aap dekh sakte hai ki filhal to humare rows ke name 01234 se hi hai par ye kuch bhi ho sakte hai or column humara ABCDE hai, "to jab humne [0,4] likha to ye 0 row or 4 column pe yaniki E par ayega to uss cell ka value hume de diya"*<br><br>***or hum isko aise bhi print kar sakte hai***

In [73]:
newdf.iloc[[0,4], [1,2]]

B         C
0  0.735723  0.351642
4  0.276021  0.140503

*to aap dekh sakte hai isne 0 or 4th row ko liya or 1 or 2 columns ko liya* ***to hum aise bhi isko print kar sakte hai***

# 10.drop
***drop se hum kisi bhi row ya column ko hata sakte hai***

In [74]:
newdf.head(3)

A         B         C         D         E
0      8521  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206

In [75]:
newdf.drop([0])

A         B         C         D         E
1    0.490249  0.464304  0.337367  0.814775  0.299901
2    0.262267  0.937619  0.831524  0.118030  0.525206
3    0.287334  0.974052  0.944044  0.869090  0.136935
4    0.840038  0.276021  0.140503  0.599835  0.511210
5    0.154071  0.032365  0.077466  0.503150  0.778425
..        ...       ...       ...       ...       ...
329  0.770738  0.692331  0.882043  0.763866  0.218966
330  0.845293  0.832767  0.497827  0.485871  0.769743
331  0.117437  0.460440  0.876562  0.636201  0.223381
332  0.381841  0.566592  0.378399  0.803676  0.020226
333  0.837817  0.830319  0.097824  0.137642  0.102835

[333 rows x 5 columns]

*to aap dekh sakte hai ki humara 0th row jo hai wo hat gaya hai yahan se*<br><br>***abb agar hume yahan se column ko hatana hai to***

In [76]:
newdf.drop(['A'], axis=1)

B         C         D         E
0    0.735723  0.351642  0.316163  0.028390
1    0.464304  0.337367  0.814775  0.299901
2    0.937619  0.831524  0.118030  0.525206
3    0.974052  0.944044  0.869090  0.136935
4    0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...
329  0.692331  0.882043  0.763866  0.218966
330  0.832767  0.497827  0.485871  0.769743
331  0.460440  0.876562  0.636201  0.223381
332  0.566592  0.378399  0.803676  0.020226
333  0.830319  0.097824  0.137642  0.102835

[334 rows x 4 columns]

*to aap dekh sakte hai humara column A yahan se gayab ho gaya hai*<br><br>***or han ye bs ek copy hai isne humare original dataframe ko change nahi kiya, humara original dataframe waisa ka waisa hi hai*** 

In [77]:
newdf.head()

A         B         C         D         E
0      8521  0.735723  0.351642  0.316163  0.028390
1  0.490249  0.464304  0.337367  0.814775  0.299901
2  0.262267  0.937619  0.831524  0.118030  0.525206
3  0.287334  0.974052  0.944044  0.869090  0.136935
4  0.840038  0.276021  0.140503  0.599835  0.511210

*aap dekh sakte hai humara original dataframe safe hai*<br> ***agar hum newdf= wo wala kar de to humara original dataframe bhi change ho jayega***

# 11.inplace
***inplace=True kar de to ye humare original dataframe ko hi modified kar deta hai***

In [78]:
newdf.drop(['A'], axis=1, inplace=True)

In [79]:
newdf

B         C         D         E
0    0.735723  0.351642  0.316163  0.028390
1    0.464304  0.337367  0.814775  0.299901
2    0.937619  0.831524  0.118030  0.525206
3    0.974052  0.944044  0.869090  0.136935
4    0.276021  0.140503  0.599835  0.511210
..        ...       ...       ...       ...
329  0.692331  0.882043  0.763866  0.218966
330  0.832767  0.497827  0.485871  0.769743
331  0.460440  0.876562  0.636201  0.223381
332  0.566592  0.378399  0.803676  0.020226
333  0.830319  0.097824  0.137642  0.102835

[334 rows x 4 columns]

***to aap dekh sakte hai isne original dataframe ko hi modify kar diya***<br>*isi tarah agar hum row ko hatayenge to*

In [80]:
newdf.drop([1,5], axis=0, inplace=True)

In [81]:
newdf

B         C         D         E
0    0.735723  0.351642  0.316163  0.028390
2    0.937619  0.831524  0.118030  0.525206
3    0.974052  0.944044  0.869090  0.136935
4    0.276021  0.140503  0.599835  0.511210
6    0.140683  0.714506  0.572959  0.167123
..        ...       ...       ...       ...
329  0.692331  0.882043  0.763866  0.218966
330  0.832767  0.497827  0.485871  0.769743
331  0.460440  0.876562  0.636201  0.223381
332  0.566592  0.378399  0.803676  0.020226
333  0.830319  0.097824  0.137642  0.102835

[332 rows x 4 columns]

*to aap dekh sakte hai hunmara 1 or 5th row jo hai yahan se hat gaya hai*

# 12.reset
**humne jo rows ko hataya abb humare index me 1 or 5 nahi hai to acha ni dikh raha hai to uss index ke number ko reset karne ke liye hum reset ka use karenge** 

In [82]:
newdf.head(3)

B         C         D         E
0  0.735723  0.351642  0.316163  0.028390
2  0.937619  0.831524  0.118030  0.525206
3  0.974052  0.944044  0.869090  0.136935

In [83]:
newdf.reset_index()

index         B         C         D         E
0        0  0.735723  0.351642  0.316163  0.028390
1        2  0.937619  0.831524  0.118030  0.525206
2        3  0.974052  0.944044  0.869090  0.136935
3        4  0.276021  0.140503  0.599835  0.511210
4        6  0.140683  0.714506  0.572959  0.167123
..     ...       ...       ...       ...       ...
327    329  0.692331  0.882043  0.763866  0.218966
328    330  0.832767  0.497827  0.485871  0.769743
329    331  0.460440  0.876562  0.636201  0.223381
330    332  0.566592  0.378399  0.803676  0.020226
331    333  0.830319  0.097824  0.137642  0.102835

[332 rows x 5 columns]

*to yahan aap dekh sakte hai humara index reset ho gaya hai* ***par yahan ye "index" name ka ek or column add kar deta hai jo hume chahiye nahi, to isko hatane ke liye hum drop=True likh denge***

In [84]:
newdf.reset_index(drop=True, inplace=True)

In [85]:
newdf.head(3)

B         C         D         E
0  0.735723  0.351642  0.316163  0.028390
1  0.937619  0.831524  0.118030  0.525206
2  0.974052  0.944044  0.869090  0.136935

*to aap dkh sakte hai yahan se index name ka column abb hat gaya hai or humne isko inplace=True bhi kar diya hai*<br>***or humara index abb reset ho gya hai or ye o se start hogi or jitni rows hai uski n-1 tak jayega. yaniki 0 se 5 tak ja raha hai matlb 6 rows hongi***

# 13.isnull
***ye yahan mujhe return karta hai wo sari values jahan jahan par ye null hai, to jahan par bhi ye True ayega wahan 0 hai***

In [86]:
newdf

B         C         D         E
0    0.735723  0.351642  0.316163  0.028390
1    0.937619  0.831524  0.118030  0.525206
2    0.974052  0.944044  0.869090  0.136935
3    0.276021  0.140503  0.599835  0.511210
4    0.140683  0.714506  0.572959  0.167123
..        ...       ...       ...       ...
327  0.692331  0.882043  0.763866  0.218966
328  0.832767  0.497827  0.485871  0.769743
329  0.460440  0.876562  0.636201  0.223381
330  0.566592  0.378399  0.803676  0.020226
331  0.830319  0.097824  0.137642  0.102835

[332 rows x 4 columns]

In [87]:
newdf['B'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
327    False
328    False
329    False
330    False
331    False
Name: B, Length: 332, dtype: bool

*to aap dekh sakte jahan jahan par ye null hai wahan isne False return kiya hai,* ***agar yahan par 0 ya none hota to ye True return karta***<br><br>
***or agar mujhe B ke sare values 0 karna hai ya None karna hai to***

In [88]:
newdf['B'] = None

In [89]:
newdf

B         C         D         E
0    None  0.351642  0.316163  0.028390
1    None  0.831524  0.118030  0.525206
2    None  0.944044  0.869090  0.136935
3    None  0.140503  0.599835  0.511210
4    None  0.714506  0.572959  0.167123
..    ...       ...       ...       ...
327  None  0.882043  0.763866  0.218966
328  None  0.497827  0.485871  0.769743
329  None  0.876562  0.636201  0.223381
330  None  0.378399  0.803676  0.020226
331  None  0.097824  0.137642  0.102835

[332 rows x 4 columns]

*to aap dekh sakte hai yahan B ke sare ke sare values None ho gaye hai*<br><br>***or aab agar hum newdf['B'] me isnull karenge to hume sare values True dikhayega*** 

In [90]:
newdf['B'].isnull()

0      True
1      True
2      True
3      True
4      True
       ... 
327    True
328    True
329    True
330    True
331    True
Name: B, Length: 332, dtype: bool

*abb ye sare hume True dikhayega*

## Note: Abb yahan par aap istarah isko kabhi na set kare, aaplog humesa .loc ki madad se hi set kare.

In [91]:
newdf.loc[:, ['B']] = None

In [92]:
newdf

B         C         D         E
0    None  0.351642  0.316163  0.028390
1    None  0.831524  0.118030  0.525206
2    None  0.944044  0.869090  0.136935
3    None  0.140503  0.599835  0.511210
4    None  0.714506  0.572959  0.167123
..    ...       ...       ...       ...
327  None  0.882043  0.763866  0.218966
328  None  0.497827  0.485871  0.769743
329  None  0.876562  0.636201  0.223381
330  None  0.378399  0.803676  0.020226
331  None  0.097824  0.137642  0.102835

[332 rows x 4 columns]

*to aap dekh sakte humne abb isko .loc ki madad se set karke humne iske value ko change kar diya*<br><br>***to aap yahan = equalto me kuch bhi likh kar run kar sakte hai or .loc hi sahi tarika hai kisi bhi value ko set karne ka***<br>**.loc hi karna cahiye nahi to hum cahining ke time pareasan ho jayenge, agar hum "direct" set karte hai to kyuki wo kabhi copy return karta hai kabhi view return karta hai humare dataframe ka isiliye**

In [93]:
newdf.loc[:, ['B']] = 56

In [94]:
newdf

B         C         D         E
0    56  0.351642  0.316163  0.028390
1    56  0.831524  0.118030  0.525206
2    56  0.944044  0.869090  0.136935
3    56  0.140503  0.599835  0.511210
4    56  0.714506  0.572959  0.167123
..   ..       ...       ...       ...
327  56  0.882043  0.763866  0.218966
328  56  0.497827  0.485871  0.769743
329  56  0.876562  0.636201  0.223381
330  56  0.378399  0.803676  0.020226
331  56  0.097824  0.137642  0.102835

[332 rows x 4 columns]

*to aap dekh sakte hai abhi 56 aa gaya hai to hum equaltoo ke age kuch bhi set kar sakte hai*

# 1.Abb hum ek naya DataFrame banayenge

# (A).drop

In [95]:
df = pd.DataFrame({"name":['Alfred', 'Batman', 'catwoman'],
                  "toy":[np.nan, 'Batmobile', 'Bullwhip'],
                  "born":[pd.NaT, pd.Timestamp("1940-04-25"),
                          pd.NaT]})

In [96]:
df.head()

name        toy       born
0    Alfred        NaN        NaT
1    Batman  Batmobile 1940-04-25
2  catwoman   Bullwhip        NaT

*abb hum is dataframe se Na ko drop karenge*

In [97]:
df.dropna()

name        toy       born
1  Batman  Batmobile 1940-04-25

*to jahan jahan na tha isne usko drop kar diya hai*

In [98]:
df.dropna(how='all')

name        toy       born
0    Alfred        NaN        NaT
1    Batman  Batmobile 1940-04-25
2  catwoman   Bullwhip        NaT

**how='all':** *means jab tak sare ke sare "nan" nahi ho jate tab tak ye row or column remove nahi karega*<br>**abb hum Batmobile ki jagah par np.nan or Bullwhip ke jagah par bhi np.nan karenge or fir how='all' ko dekhenge**

In [99]:
df = pd.DataFrame({"name":['Alfred', 'Batman', 'catwoman'],
                  "toy":[np.nan, np.nan, np.nan],
                  "born":[pd.NaT, pd.Timestamp("1940-04-25"),
                          pd.NaT]})

In [100]:
df

name  toy       born
0    Alfred  NaN        NaT
1    Batman  NaN 1940-04-25
2  catwoman  NaN        NaT

In [101]:
df.dropna(how='all', axis=1)

name       born
0    Alfred        NaT
1    Batman 1940-04-25
2  catwoman        NaT

*to abb aap dekh sakte hai humara toy column gayab ho chuka hai, or humne yahan axis=1 isiliye kiya kyuki axis=1 means column hota hai*

# (B).drop_duplicates

In [102]:
df = pd.DataFrame({"name":['Alfred', 'Batman', 'Alfred'],
                  "toy":[np.nan, 'Batmobile', 'Bullwhip'],
                  "born":[pd.NaT, pd.Timestamp("1940-04-25"),
                          pd.NaT]})

In [103]:
df

name        toy       born
0  Alfred        NaN        NaT
1  Batman  Batmobile 1940-04-25
2  Alfred   Bullwhip        NaT

*to humne yahan par catwoman ko hata kar Alfred kar diya hai jisse name me duplicate hojayega*

In [104]:
df.drop_duplicates(subset=['name'])

name        toy       born
0  Alfred        NaN        NaT
1  Batman  Batmobile 1940-04-25

*to isne yahan se duplicates ko hata diya hai*<br>**abb hum yaha first duplicate ya last duplicate rakh sakte hai ya duplicate ko puri tarah hata bhi sakte hai**

In [105]:
df.drop_duplicates(subset=['name'], keep='first')

name        toy       born
0  Alfred        NaN        NaT
1  Batman  Batmobile 1940-04-25

*isse pehla duplicate rahega*

In [106]:
df.drop_duplicates(subset=['name'], keep='last')

name        toy       born
1  Batman  Batmobile 1940-04-25
2  Alfred   Bullwhip        NaT

*isse last duplicate rahega*

In [107]:
df.drop_duplicates(subset=['name'], keep=False)

name        toy       born
1  Batman  Batmobile 1940-04-25

*or isse duplicate hi hat jayega*

# ()Some Extra Functions
***agar apko ek data set diya gaya hai to usko inspect kis tarah karna hai***

In [108]:
df.head()

name        toy       born
0  Alfred        NaN        NaT
1  Batman  Batmobile 1940-04-25
2  Alfred   Bullwhip        NaT

In [109]:
df.tail()

name        toy       born
0  Alfred        NaN        NaT
1  Batman  Batmobile 1940-04-25
2  Alfred   Bullwhip        NaT

In [110]:
df.describe()

C:\Users\SANJIV~1\AppData\Local\Temp/ipykernel_11692/3627053830.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df.describe()


name        toy                 born
count        3          2                    1
unique       2          2                    1
top     Alfred  Batmobile  1940-04-25 00:00:00
freq         2          1                    1
first      NaN        NaN  1940-04-25 00:00:00
last       NaN        NaN  1940-04-25 00:00:00

In [111]:
df.shape

(3, 3)

In [112]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   name    3 non-null      object        
 1   toy     2 non-null      object        
 2   born    1 non-null      datetime64[ns]
dtypes: datetime64[ns](1), object(2)
memory usage: 200.0+ bytes


## For count unique values

In [113]:
df['name'].value_counts(dropna=False)

Alfred    2
Batman    1
Name: name, dtype: int64

*means na ko mat hatao mt gino*

In [114]:
df['toy'].value_counts(dropna=False)

NaN          1
Batmobile    1
Bullwhip     1
Name: toy, dtype: int64

In [115]:
df['toy'].value_counts(dropna=True)

Batmobile    1
Bullwhip     1
Name: toy, dtype: int64

In [116]:
df.notnull()

name    toy   born
0  True  False  False
1  True   True   True
2  True   True  False

*iska mtlb jahan jahan null nahi hai wahan True hoga, or jahan null hai wahan False ho jayega*

# Question:

### Create a dataframe which contains only integers with 3 row and 2 columns and run following dataframe methods on them:
df.describe()<br>
df.mean()<br>
df.corr()<br>
df.count()<br>
df.max()<br>
df.min()<br>
df.median()<br>
df.std()<br>


# Answer:

In [117]:
# newdf2 = pd.DataFrame(np.random.rand(3, 2), index=np.arange(3))

In [118]:
# newdf2

In [119]:
newdf2 = pd.DataFrame(np.random.randint(0,50,size=(3, 2)), index=np.arange(3))

**Here - np.random.randint(0,100,size=(100, 4)) - creates an output array of size (100,4) with random integer elements between [0,100)**

In [120]:
newdf2

0   1
0   3   6
1  30  25
2  23   2

In [121]:
newdf2.describe()

0          1
count   3.000000   3.000000
mean   18.666667  11.000000
std    14.011900  12.288206
min     3.000000   2.000000
25%    13.000000   4.000000
50%    23.000000   6.000000
75%    26.500000  15.500000
max    30.000000  25.000000

In [122]:
newdf2.mean()

0    18.666667
1    11.000000
dtype: float64

In [123]:
newdf2.corr()

0         1
0  1.000000  0.574976
1  0.574976  1.000000

In [124]:
newdf2.count()

0    3
1    3
dtype: int64

In [125]:
newdf2.max()

0    30
1    25
dtype: int32

In [126]:
newdf2.min()

0    3
1    2
dtype: int32

In [127]:
newdf2.median()

0    23.0
1     6.0
dtype: float64

In [128]:
newdf2.std()

0    14.011900
1    12.288206
dtype: float64

# pandas.read_excel
## 1.read

In [129]:
data = pd.read_excel('data.xlsx')

In [130]:
data

train no2  speed2       city2
0         37      92  jamshedpur
1       5847     100      ranchi
2       4455      90    musabani

*to isne humari pehli wali sheet ko read kar diya* ***abb hum dushri sheet ko read karenge***

In [131]:
data = pd.read_excel('data.xlsx', sheet_name='Sheet1')

In [132]:
data

train no2  speed2       city2
0         37      92  jamshedpur
1       5847     100      ranchi
2       4455      90    musabani

*yahan humne sheet1 ko read kiya*

In [133]:
# data = pd.read_excel('data.xlsx', sheet_name='Sheet2')

In [134]:
data

train no2  speed2       city2
0         37      92  jamshedpur
1       5847     100      ranchi
2       4455      90    musabani

*or yahan humne sheet2 ko read kiya*<br> ***is tarah hum multiple sheet me read kar sakta hun or write bhi kar sakta hun***

## 2.write

In [135]:
data

train no2  speed2       city2
0         37      92  jamshedpur
1       5847     100      ranchi
2       4455      90    musabani

In [136]:
data.iloc[0,0]=37

In [137]:
data

train no2  speed2       city2
0         37      92  jamshedpur
1       5847     100      ranchi
2       4455      90    musabani

In [138]:
data.to_excel('data.xlsx', sheet_name='Sheet1')
# data2.to_excel('data.xlsx', sheet_name='Sheet2')

***isne write to kar diya lekin jo humare sheet 2 ko delete kar diya isne to humne isko or ache se dekhne ka zarrorat hai***<br>**note: yahan par humare sheet ke name ka S capital hai to humne isko capital liya hai apka name jaisa ho waisa hi le aap**

# Note:
***google search kare or official documentation dekhe pandas ka***<br>
**search like = "pandas.ExcelWriter"**

# Thank you